examples/all-clusters-app/linux/diagnostic-logs-provider-delegate-impl.cpp

```diff
@@ -0,0 +1,181 @@
+
+LogProvider LogProvider::sInstance;
+
+namespace {
+bool IsValidIntent(IntentEnum intent)
+{
+    return intent == IntentEnum::kEndUserSupport || intent == IntentEnum::kNetworkDiag || intent == IntentEnum::kCrashLogs;
+}
+
+size_t GetFileSize(FILE * fp)
+{
+    VerifyOrReturnValue(nullptr != fp, 0);
+
+    // Get the current position.
+    auto offset = ftell(fp);
+
+    // Move to the end.
+    fseek(fp, 0, SEEK_END);
+
+    // Read the size.
+    auto fileSize = ftell(fp);
+
+    // Move back to the original position.
+    fseek(fp, offset, SEEK_SET);
+
+    return static_cast<size_t>(fileSize);
+}
+} // namespace
+
+CHIP_ERROR LogProvider::GetLogForIntent(chip::app::Clusters::DiagnosticLogs::IntentEnum intent, chip::MutableByteSpan & outBuffer,
+                                        const chip::Optional<uint64_t> & outTimeStamp,
+                                        const chip::Optional<uint64_t> & outTimeSinceBoot)
+{
+    CHIP_ERROR err                 = CHIP_NO_ERROR;
+    LogSessionHandle sessionHandle = kInvalidLogSessionHandle;
+
+    err = StartLogCollection(intent, sessionHandle, outTimeStamp, outTimeSinceBoot);
+    if (CHIP_NO_ERROR != err)
+    {
+        outBuffer.reduce_size(0);
+        return err;
+    }
+
+    bool outIsEndOfLog;
+    err = CollectLog(sessionHandle, outBuffer, outIsEndOfLog /* unused */);
+    if (CHIP_NO_ERROR != err)
+    {
+        outBuffer.reduce_size(0);
+        return err;
+    }
+
+    err = EndLogCollection(sessionHandle);
+    if (CHIP_NO_ERROR != err)
+    {
+        outBuffer.reduce_size(0);
+        return err;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR LogProvider::StartLogCollection(IntentEnum intent, LogSessionHandle & outHandle,
+                                           const chip::Optional<uint64_t> & outTimeStamp,
+                                           const chip::Optional<uint64_t> & outTimeSinceBoot)
+{
+    VerifyOrReturnValue(IsValidIntent(intent), CHIP_ERROR_INVALID_ARGUMENT);
+
+    auto filePath = GetFilePathForIntent(intent);
+    VerifyOrReturnValue(filePath.HasValue(), CHIP_ERROR_NOT_FOUND);
+
+    auto fp = fopen(filePath.Value().c_str(), "rb");
+    VerifyOrReturnValue(nullptr != fp && errno != ENOENT, CHIP_ERROR_NOT_FOUND);
+    VerifyOrReturnValue(nullptr != fp, CHIP_ERROR_INTERNAL);
+
+    mLogSessionHandle++;
+    // If the session handle rolls over to UINT16_MAX which is invalid, reset to 0.
+    VerifyOrDo(mLogSessionHandle != kInvalidLogSessionHandle, mLogSessionHandle = 0);
+
+    outHandle                 = mLogSessionHandle;
+    mFiles[mLogSessionHandle] = fp;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR LogProvider::EndLogCollection(LogSessionHandle sessionHandle)
+{
+    VerifyOrReturnValue(sessionHandle != kInvalidLogSessionHandle, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnValue(mFiles.count(sessionHandle), CHIP_ERROR_INVALID_ARGUMENT);
+
+    auto fp = mFiles[sessionHandle];
+    fclose(fp);
```

Comment:
after close, should we set the session handle null/clear it?

Reply:
fclose may fail. Please check for return value of any methods that may fail

Reply:
I have cleared the session handle. Good catch.